test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/TimeoutMock.scala

```diff
@@ -0,0 +1,110 @@
+package org.scalajs.testsuite.jsinterop
+
+import scala.annotation.tailrec
+import scala.collection.immutable.TreeMap
+import scala.scalajs.js
+import scala.scalajs.js.annotation.ScalaJSDefined
+
+object TimeoutMock {
+
+  import js.Dynamic.global
+
+  private val realSetTimeout =
+    global.setTimeout.asInstanceOf[js.Function]
+
+  private val realClearTimeout =
+    global.clearTimeout.asInstanceOf[js.Function]
+
+  private val mockSetTimeout: js.Function =
+    (fun: js.Function, delay: Int) => MockTimeouts.setTimeout(fun, delay)
+
+  private val mockClearTimeout: js.Function =
+    (timeout: MockTimeout) => MockTimeouts.clearTimeout(timeout)
+
+  private var installed = false
+
+  @noinline
+  def withMockedTimeout[A](body: (Int => Unit) => A): A = {
+    install()
+    try {
+      body(MockTimeouts.tick)
+    } finally {
+      uninstall()
+    }
+  }
+
+  private def install(): Unit = {
+    assert(!installed, "Mock timeout already installed.")
+    global.setTimeout = mockSetTimeout
+    global.clearTimeout = mockClearTimeout
+    installed = true
+  }
+
+  private def uninstall(): Unit = {
+    assert(installed, "Mock timeout is not installed.")
+    MockTimeouts.reinitialize()
+    global.setTimeout = realSetTimeout
+    global.clearTimeout = realClearTimeout
+    installed = false
+  }
+
+  private object MockTimeouts {
+    private var timeouts: TreeMap[Int, List[MockTimeout]] = _
```

Comment:
There is not `mutable.TreeMap` in `scala.collection` and no other map that offers the complexity characteristics required here.